server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -276,10 +280,12 @@ private static class WeightFunction {
         }
 
         float weight(Balancer balancer, ModelNode node, String index) {
-            final float weightShard = node.numShards() - balancer.avgShardsPerNode();
+            final var tier = balancer.indexTier(index);
+            final var tierStats = balancer.getStatsForTier(tier);
+            final float weightShard = node.numShards() - tierStats.avgShardSize();
```

Comment:
I originally confused `tierStats.avgShardSize()` with `avgDiskUsageInBytes`. 
Should we rename that to `avgShardsCount`?